contracts/governance/rewards/RewardsDistributor.sol

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import {ReentrancyGuard} from '@openzeppelin/contracts/utils/ReentrancyGuard.sol';
+import {SafeMath} from '@openzeppelin/contracts/math/SafeMath.sol';
+import {SafeERC20} from '@openzeppelin/contracts/token/ERC20/SafeERC20.sol';
+import {IERC20Ext} from "@kyber.network/utils-sc/contracts/IERC20Ext.sol";
+import {Utils} from '@kyber.network/utils-sc/contracts/Utils.sol';
+import {PermissionAdmin} from '@kyber.network/utils-sc/contracts/PermissionAdmin.sol';
+import {MerkleProof} from '../../misc/MerkleProof.sol';
+import {IPool} from '../../interfaces/IPool.sol';
+
+contract RewardsDistributor is PermissionAdmin, ReentrancyGuard, Utils {
+    
+    using SafeERC20 for IERC20Ext;
+    using SafeMath for uint256;
+
+    struct MerkleData {
+        uint256 cycle;
+        bytes32 root;
+        string contentHash;
+    }
+    
+    IPool public treasuryPool;
+    MerkleData private merkleData;
+    // wallet => token => claimedAmount
+    mapping(address => mapping(IERC20Ext => uint256)) public claimedAmounts;
+
+    event TreasuryPoolSet(IPool indexed treasuryPool);
+    event Claimed(uint256 indexed cycle, address indexed user, IERC20Ext token, uint256 claimAmount);
+    event RootUpdated(uint256 indexed cycle, bytes32 root, string contentHash);
+
+    constructor(address admin, IPool _treasuryPool) PermissionAdmin(admin) {
+        treasuryPool = _treasuryPool;
+    }
+
+    receive() external payable {}
+
+    function getMerkleData() external view returns (MerkleData memory) {
+        return merkleData;
+    }
+
+    function isValidClaim(
+        uint256 cycle,
+        uint256 index,
+        address user,
+        IERC20Ext[] calldata tokens,
+        uint256[] calldata cumulativeAmounts,
+        bytes32[] calldata merkleProof
+    ) external view returns (bool) {
+        if (cycle != merkleData.cycle) return false;
+        bytes32 node = keccak256(abi.encodePacked(cycle, index, user, tokens, cumulativeAmounts));
+        return MerkleProof.verify(merkleProof, merkleData.root, node);
+    }
+
+    /// @notice Claim accumulated rewards for a set of tokens at a given cycle number
+    function claim(
+        uint256 cycle,
+        uint256 index,
+        IERC20Ext[] calldata tokens,
+        uint256[] calldata cumulativeAmounts,
+        bytes32[] calldata merkleProof
+    ) external nonReentrant returns (uint256[] memory claimAmounts) {
+        require(cycle == merkleData.cycle, 'incorrect cycle');
+
+        // verify the merkle proof
+        bytes32 node = keccak256(abi.encodePacked(cycle, index, msg.sender, tokens, cumulativeAmounts));
```

Comment:
For flexibily, I prefer user can have a field `account: address` to claim.

For example, if user want to claim reward for a smart-contract address.